ride/referral.ride

```diff
@@ -0,0 +1,197 @@
+
+  ([
+    BooleanEntry(keyExistsReferrerToReferral(programName, referrerAddress, referralAddress), true),
+    IntegerEntry(keyTotalReferralCount(programName, referrerAddress), newTotalReferralCount)
+  ], unit)
+}
+
+@Callable(i)
+func calculateActiveReferralCount(
+  programName: String,
+  referrerAddress: String,
+  activeReferralCount: Int
+) = {
+  let incAllowedContract = getString(keyProgramAllowedContract(programName)).value()
+  strict checkCaller = i.caller.toString() == incAllowedContract || throw("Permission denied")
+
+  ([IntegerEntry(keyActiveReferralCount(programName, referrerAddress), activeReferralCount)], unit)
+}
+
+@Callable(i)
+func incUnclaimed(
+  programName: String,
+  referralAddress: String,
+  referrerReward: Int,
+  referralReward: Int
+) = {
+  let incAllowedContract = getString(keyProgramAllowedContract(programName)).value()
+  strict checkCaller = i.caller.toString() == incAllowedContract || throw("Permission denied")
+
+  let referrerAddress = getString(keyReferrer(programName, referralAddress)).value()
+  let referrerUnclaimed = getInteger(keyUnclaimed(programName, referrerAddress)).value()
+  let referralUnclaimed = getInteger(keyUnclaimed(programName, referrerAddress)).value()
+
+  let newReferrerUnclaimed = referrerUnclaimed + referrerReward
+  let newReferralUnclaimed = referralUnclaimed + referralReward
+
+  ([
+    IntegerEntry(keyClaimed(programName, referralAddress), newReferrerUnclaimed),
+    IntegerEntry(keyUnclaimed(programName, referralAddress), newReferralUnclaimed)
+  ], unit)
+}
+
+@Callable(i)
+func claim(programName: String) = {
+  let claimedTotal = getInteger(keyClaimedTotal(programName)).value()
+  let claimerClaimed = getInteger(keyClaimed(programName, i.caller.toString())).value()
+  let claimerUnclaimed = getInteger(keyUnclaimed(programName, i.caller.toString())).value()
+  
+  strict checkCanClaim = claimerUnclaimed > 0 || "nothing to claim".throw()
+
+  let newClaimerClaimed = claimerClaimed + claimerUnclaimed
+  let newClaimedTotal = claimedTotal + claimerUnclaimed
+  let wxAssetId = getString(keyWxAssetId).value().fromBase58String()
+
+  strict balanceBefore = assetBalance(this, wxAssetId).value()
+  strict res = marketingContract.invoke("withdrawReferralReward", [claimerUnclaimed], [])
+  strict balanceAfter = assetBalance(this, wxAssetId).value()
+  strict balanceDiff = balanceAfter - balanceBefore
+  strict checkBalance = balanceDiff == claimerUnclaimed || "Insufficient balance on referral contract".throw()
+
+  ([
+    IntegerEntry(keyClaimed(programName, i.caller.toString()), newClaimerClaimed),
+    IntegerEntry(keyClaimedTotal(programName), newClaimedTotal),
+    IntegerEntry(keyUnclaimed(programName, i.caller.toString()), 0),
+    ScriptTransfer(i.caller, claimerUnclaimed, wxAssetId)
+  ], unit)
+}
+
+@Callable(i)
+func claimREADONLY(programName: String) = {
+  let claimerClaimed = getInteger(keyClaimed(programName, i.caller.toString())).value()
```

Comment:
It's impossible to use the function because there's no caller in `/utils/evaluate`